### ISSUE TITLE
Should user-defined types be able to used in conditionals?  And if so, how?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#17645]
As a library author, I may want to define a type that would be convenient to use in a conditional.  As a somewhat artificial example, if I define a type:

```chapel
record height {
  var inches: int; 
}
```

I may want to permit a user to write:

```chapel
if child.height then
```

rather than:

```chapel
if child.height.inches > 0 then  // or `if child.height.getHeightInInches() > 0` or ...
```

In the modules developed by the core team, we have benefitted from this capability in the `Regex` module, for `c_ptr`, and for the `syserr` type, all of which are defined within library modules.  Should we extend this capability to users as well?

Pros:
* supports creating the ability to create types with succinct conditional forms

Cons:
* like operator overloads, could be subject to abuse
* somewhat similar to an implicit conversion, which some users have been averse to due to things happening invisibly

If we were to support this, how would we express it?

* a well-defined method on the type?
* a (new) operator on the type?

In either case, what would the name be?

Related issues:
* https://github.com/chapel-lang/chapel/issues/14930
* https://github.com/chapel-lang/chapel/issues/15904
* https://github.com/chapel-lang/chapel/issues/17633

**Comment by bradcray** [comment]
> It's unclear to me what it could be as an operator - what would a "for instance" idea there be?

The only idea I've come up with so far is `operator if()`, which is a bit weird except that (a) there are no naming conflicts and (b) it makes it clear that it's for use in conditionals (but it's also for use in the condition expressions in while loops which makes it a bit weird).

Spitballing further:

I don't have any intuition as to whether this would make more sense as an operator than a procedure.  The main argument I have for "operator" is "it's implementing something that provides a language capability."  The main argument I have for procedure is "it's not invoked using an operator symbol" (nor defined using one, so far).

I also wonder whether—if we had user-defined coercions—we'd want to implement this capability by defining the "coerce-to-bool" overload and its presence would cause it to be used here.  Or would we want those to be two separate things (you can coerce to bool but not use it in a conditional — or vice-versa).  My intuition is "no, you should get both of these things or neither."  So this might suggest defining a user-defined coercion operator?

Or, given that we have casts, should we define this using a cast-to-bool case:  `operator :(x: mytype, type b: bool) { ... }` and say that the compiler compiles `if expr` and `while expr` into `if expr:bool` and `while expr:bool`?  Raises the same question as just above:  Would we want those to be two separate things?  (i.e., I want my type to support casts to bool but not to be usable in conditionals; or I want my type to be usable in conditionals but not to be cast to bools).  This feels like a less obvious "no" than the previous.

One thing I realized today is that even if we didn't give users this capability, we already effectively do because they can define the `!` operator.  That is, if I can't say `if myType then...`, I can define `!` for `myType` and then say `if !myType then ...` which seems unfair / asymmetric.  If we didn't link this to casts or coercions, that takes me to the idea of defining it as defining it as `operator !!`.  Or the compiler could convert all conditionals `if myExpr` into `if !(!myExpr)` which wouldn't require us to introduce any new things into the language, but would have the slightly silly property of relying on a double negative to implement every conditional.

Engin's reaction here:

>  am used to if match from Python, too. But I wouldn't lose any sleep if we were to drop that support.

Is why I think we should support this, combined with the "`!` already gives this capability" observation above.  I don't think we're doing anything more dangerous / open to abuse than we already have while enabling users to create their own patterns that are as convenient as the reMatch case (surely there are other patterns like that which Engin already relies on without realizing it?)

**Comment by mppf** [comment]
> The main argument I have for "operator" is "it's implementing something that provides a language capability."

> > Relatedly, if this one were to be an operator, should we consider changing this and these into operators?

I'd like to discuss this more but it's starting to get off-topic. Taken to an extreme, we'd also write `operator init=` and `operator init`.

> I also wonder whether—if we had user-defined coercions—we'd want to implement this capability by defining the "coerce-to-bool" overload and its presence would cause it to be used here. Or would we want those to be two separate things (you can coerce to bool but not use it in a conditional — or vice-versa). My intuition is "no, you should get both of these things or neither." So this might suggest defining a user-defined coercion operator?

One of the key points from the "levels of conversions" idea is that there is a different sort of risk of confusion for conversions depending on the situation. Implicit conversions for a function call having the most potential to create confusion. In this case specifically, the conversion to `bool` for `if myRecord` does not have those potential negative impacts related to not knowing which function is called. So I think that it should be something that could be requested separately. (In other words, for conversion to `bool` specifically, we should be able to choose a) the type implicitly converts to `bool` (and can be used in an `if`) or b) the type can be used in `if` but can't be implicitly converted to `bool` for a function call or c) the type has no particular relationship with `bool`).

So this leads me to thinking that we do need a separate way to request this. But, none of that is an objection to the "if x always casts x to bool" strategy.

The point about `!!` is interesting. I have seen C++ code that uses `if (!!x)` which I really don't like but I suppose it is solving a problem in C++. I'd like us to not have that problem. But, using `!!` as the name seems OK.

Seems like we have two choices:
 * should it be introduced by `proc` or `operator` ?
 * what is the name of the function/method/operator to overload?
    * `if`
    * `toCondition`
    * `!!`
    * `:` (i.e. it is the same as defining cast to bool)

**Comment by e-kayrakli** [comment]
> I don't think we're doing anything more dangerous / open to abuse than we already have while enabling users to create their own patterns that are as convenient as the reMatch case (surely there are other patterns like that which Engin already relies on without realizing it?)

I agree with this in the sense that, if we keep the support we have for regexMatch type, then it should be via a feature that's user facing. Also, good point about `operator !` almost doing the job in an indirect way.

The general idea (including the support for regexMatch) is still something that I am not psyched about. But I don't have any other concrete arguments against it besides my paranoia about operator overloads.

**Comment by bradcray** [comment]
Mentioning `_cond_test` here simply because in searching for it, I didn't come across this issue (_cond_test is how we express such things internally today).
